ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1527 @@
 	}
-	identd->filter_list_size = num;
-
-	/* set strict flag */
+	identd->filter_list_size = elistsz;
 	identd->strict = strict;
 
 	return identd;
 }
 
-
-void
-mbfl_encoding_detector_delete(mbfl_encoding_detector *identd)
+void mbfl_encoding_detector_delete(mbfl_encoding_detector *identd)
 {
-	int i;
-
-	if (identd != NULL) {
-		if (identd->filter_list != NULL) {
-			i = identd->filter_list_size;
-			while (i > 0) {
-				i--;
-				mbfl_identify_filter_delete(identd->filter_list[i]);
-			}
-			efree((void *)identd->filter_list);
-		}
-		efree((void *)identd);
+	for (int i = identd->filter_list_size - 1; i >= 0; i--) {
+		mbfl_identify_filter_delete(identd->filter_list[i]);
 	}
+	efree((void*)identd->filter_list);
+	efree((void*)identd);
 }
 
-int
-mbfl_encoding_detector_feed(mbfl_encoding_detector *identd, mbfl_string *string)
+int mbfl_encoding_detector_feed(mbfl_encoding_detector *identd, mbfl_string *string)
 {
-	int res = 0;
-	/* feed data */
-	if (identd != NULL && string != NULL && string->val != NULL) {
+	if (string->len > 0) {
 		int num = identd->filter_list_size;
-		size_t n = string->len;
-		unsigned char *p = string->val;
+		unsigned char *p = string->val, *e = p + string->len;
 		int bad = 0;
-		while (n > 0) {
-			int i;
-			for (i = 0; i < num; i++) {
+
+		/* We have an array of filter functions corresponding to different encodings
+		 * Iterate over the string, pushing each byte into each and every filter function
+		 *
+		 * If the string is not encoding XYZ, then likely at some point we will hit a
+		 * byte which is not valid in encoding XYZ, and the the 'flag' for that encoding
+		 * will be set.
+		 *
+		 * Our hope is that at some point, all the candidate encodings will be flagged
+		 * except one, which will presumably be the correct one. */
+		while (p < e) {
+			for (int i = 0; i < num; i++) {
 				mbfl_identify_filter *filter = identd->filter_list[i];
 				if (!filter->flag) {
-					(*filter->filter_function)(*p, filter);
+					(filter->filter_function)(*p, filter);
 					if (filter->flag) {
 						bad++;
 					}
 				}
 			}
-			if ((num - 1) <= bad) {
-				res = 1;
-				break;
+			/* If all candidate encodings have failed, or there is only one remaining
+			 * which has not failed, we can return.
+			 *
+			 * The exception is in strict mode; because when we discover that all
+			 * other possible encodings have failed, we might still be in the middle
+			 * of a multi-byte character for the 'correct' encoding, and stopping
+			 * right here would cause `mbfl_encoding_detector_judge` (below) to
+			 * wrongly reject it. */
+			if ((num - 1) <= bad && !identd->strict) {
+				return 1;
 			}
 			p++;
-			n--;
 		}
 	}
 
-	return res;
+	return 0; /* No conclusion on which of the candidate encodings is correct */
 }
 
-const mbfl_encoding *mbfl_encoding_detector_judge(mbfl_encoding_detector *identd)
+const mbfl_encoding* mbfl_encoding_detector_judge(mbfl_encoding_detector *identd)
```

Comment:
```suggestion
const mbfl_encoding *mbfl_encoding_detector_judge(mbfl_encoding_detector *identd)
```
This one got changed the wrong way around :)
Really applies to all functions...

Reply:
Fixed all of them.